contracts/fee_currencies.go

```diff
@@ -0,0 +1,119 @@
+package contracts
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/celo/abigen"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	Thousand = 1000
+	Million  = 1000 * 1000
+
+	maxGasForDebitGasFeesTransactions  uint64 = 1 * Million
+	maxGasForCreditGasFeesTransactions uint64 = 1 * Million
+	// Default intrinsic gas cost of transactions paying for gas in alternative currencies.
+	// Calculated to estimate 1 balance read, 1 debit, and 4 credit transactions.
+	IntrinsicGasForAlternativeFeeCurrency uint64 = 50 * Thousand
+)
+
+var (
+	tmpAddress = common.HexToAddress("0xce106a5")
+)
+
+// GetBalanceOf returns an account's balance on a given ERC20 currency
+func GetBalanceOf(caller bind.ContractCaller, accountOwner common.Address, contractAddress common.Address) (result *big.Int, err error) {
+	token, err := abigen.NewFeeCurrencyCaller(contractAddress, caller)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access FeeCurrency: %w", err)
+	}
+
+	balance, err := token.BalanceOf(&bind.CallOpts{}, accountOwner)
+	if err != nil {
+		return nil, err
+	}
+	log.Trace("GetBalanceOf() Called", "accountOwner", accountOwner.Hex(), "contractAddress", contractAddress)
+
+	return balance, nil
+}
+
+func ConvertGoldToCurrency(exchangeRates map[common.Address]*big.Rat, feeCurrency *common.Address, goldAmount *big.Int) *big.Int {
+	er := exchangeRates[*feeCurrency]
+	return new(big.Int).Div(new(big.Int).Mul(goldAmount, er.Num()), er.Denom())
+}
+
+func DebitFees(evm *vm.EVM, address common.Address, amount *big.Int, feeCurrency *common.Address) error {
+	if amount.Cmp(big.NewInt(0)) == 0 {
+		return nil
+	}
+	abi, err := abigen.FeeCurrencyMetaData.GetAbi()
+	if err != nil {
+		return err
+	}
+	// Solidity: function transfer(address to, uint256 amount) returns(bool)
+	input, err := abi.Pack("transfer", tmpAddress, amount)
+	if err != nil {
+		return err
+	}
+
+	caller := vm.AccountRef(address)
+
+	_, leftoverGas, err := evm.Call(caller, *feeCurrency, input, maxGasForDebitGasFeesTransactions, big.NewInt(0))
+	gasUsed := maxGasForDebitGasFeesTransactions - leftoverGas
+	fmt.Println("DebitFeesNew called", "feeCurrency", *feeCurrency, "gasUsed", gasUsed)
+	return err
+}
+
+func CreditFees(
+	evm *vm.EVM,
+	from common.Address,
+	feeRecipient common.Address,
+	feeHandler common.Address,
+	refund *big.Int,
+	tipTxFee *big.Int,
+	baseTxFee *big.Int,
+	feeCurrency *common.Address) error {
+	caller := vm.AccountRef(tmpAddress)
+
+	abi, err := abigen.FeeCurrencyMetaData.GetAbi()
+	if err != nil {
+		return err
+	}
+
+	// Solidity: function transfer(address to, uint256 amount) returns(bool)
+	transfer1Data, err := abi.Pack("transfer", from, refund)
```

Comment:
probably we could check if the refund is zero to avoid this call. Same with the tip and base (which shouldn't be zero as far as I know, but that check could save some process)

Reply:
Done in 47192a5f12edfb87975d042d39a8e4c651e70afa

I left out the check for the base fee, as that should never be zero for us.